### PR TITLE
Filter referrers by artifact type

### LINKED ISSUE #140: Implement `aritfactType` filtering
The current [remote.Repository](https://github.com/oras-project/oras-go/blob/main/registry/remote/repository.go#L343) does not support [artifact type filtering](https://github.com/oras-project/artifacts-spec/blob/main/manifest-referrers-api.md#filtering-results) for the referral API. It is marked as an to-do item.

### PR DESCRIPTION
**Pull request by m5i-work**
Resolves #140 

Tested against oras-distribution and ACR DF3

Signed-off-by: Wei Meng <wemeng@microsoft.com>

### CODE REVIEW

File: registry/remote/repository.go [hunk 1]
```diff
@@ -407,13 +408,35 @@ func (r *Repository) referrers(ctx context.Context, desc ocispec.Descriptor, fn
...
 }
 
+// filterReferrers filters a slice of referrers by artifactType in place.
+// The returned slice contains matching referrers.
+func filterReferrers(refs []artifactspec.Descriptor, artifactType string) []artifactspec.Descriptor {
+	if artifactType == "" {
+		return refs
+	}
+	var o int
+	for i := 0; i < len(refs); i++ {
```
Comment (Wwwsylvia) [inline registry/remote/repository.go:429]:
```suggestion
	for i, ref := range refs {
```

File: registry/remote/repository.go [hunk 2]
```diff
@@ -407,13 +408,35 @@ func (r *Repository) referrers(ctx context.Context, desc ocispec.Descriptor, fn
...
 	return parseLink(resp)
 }
 
+// filterReferrers filters a slice of referrers by artifactType in place.
+// The returned slice contains matching referrers.
+func filterReferrers(refs []artifactspec.Descriptor, artifactType string) []artifactspec.Descriptor {
+	if artifactType == "" {
+		return refs
+	}
+	var o int
```
Comment (Wwwsylvia) [inline registry/remote/repository.go:428]: Just curious, why is this called `o`?

File: registry/remote/repository_test.go [hunk 3]
```diff
@@ -1146,6 +1146,335 @@ func TestRepository_Referrers_Fallback(t *testing.T) {
...
+			if q.Get("digest") == "" {
+				t.Errorf("digest not provided: %s %q", r.Method, r.URL)
+				w.WriteHeader(http.StatusBadRequest)
+				return
+			}
+			referrers = referrerSet[0]
+			w.Header().Set("Link", fmt.Sprintf(`<%s?n=2&test=foo>; rel="next"`, path))
+		}
+		result := struct {
+			References []artifactspec.Descriptor `json:"references"`
```
Comment (Wwwsylvia) [inline registry/remote/repository_test.go:1229]: Shouldn't this be `Referrers`?

File: registry/remote/repository.go (hunk 2 above)
Comment (shizhMSFT) [reply registry/remote/repository.go:428]: may be `i` and `j` are better names?

**Review by shizhMSFT — APPROVED** [review]
LGTM

**Review by Wwwsylvia — APPROVED** [review]
LGTM

File: registry/remote/url.go [hunk 4]
```diff
@@ -87,25 +88,39 @@ func buildRepositoryBlobUploadURL(plainHTTP bool, ref registry.Reference) string
...
 
 // buildArtifactReferrerURLLegacy builds the URL for accessing the manifest referrers API in artifact spec v1.0.0-draft.1.
-// Format: <scheme>://<registry>/oras/artifacts/v1/<repository>/manifests/<digest>/referrers
+// Format: <scheme>://<registry>/oras/artifacts/v1/<repository>/manifests/<digest>/referrers?artifactType=<artifactType>
 // Reference: https://github.com/oras-project/artifacts-spec/blob/v1.0.0-draft.1/manifest-referrers-api.md
-func buildArtifactReferrerURLLegacy(plainHTTP bool, ref registry.Reference) string {
+func buildArtifactReferrerURLLegacy(plainHTTP bool, ref registry.Reference, artifactType string) string {
+	var query string
+	if artifactType != "" {
+		v := url.Values{}
```
[the remark this answers is not in the document]
Comment (shizhMSFT) [reply registry/remote/url.go:96]: That makes sense.

**Review by shizhMSFT — APPROVED** [review]
LGTM

File: registry/remote/repository.go [hunk 5]
```diff
@@ -407,13 +408,35 @@ func (r *Repository) referrers(ctx context.Context, desc ocispec.Descriptor, fn
...
+	var j int
+	for i, ref := range refs {
+		if ref.ArtifactType == artifactType {
+			if i != j {
+				refs[j] = ref
+			}
+			j++
+		}
+	}
+	return refs[:j]
```
Comment (qweeah) [inline registry/remote/repository.go:437]: Cool in-place filtering!

**Pull request merged**